autometrics/src/alerts.rs

```diff
@@ -0,0 +1,295 @@
+// The alert definitions are collected into this slice at compile time.
+// See https://github.com/dtolnay/linkme for how this works.
+#[doc(hidden)]
+#[distributed_slice]
+pub static METRICS: [Alert] = [..];
+
+#[doc(hidden)]
+pub struct Alert {
+    pub function: &'static str,
+    pub module: &'static str,
+    pub success_rate: Option<&'static str>,
+    pub latency: Option<(&'static str, &'static str)>,
+}
+
+impl Alert {
+    fn to_objectives(&self) -> impl Iterator<Item = Box<dyn Objective>> {
+        let mut objectives: Vec<Box<dyn Objective>> = vec![];
+        if let Some(success_rate) = self.success_rate {
+            objectives.push(Box::new(SuccessRateObjective {
+                function: self.function,
+                module: self.module,
+                success_rate,
+            }));
+        }
+        if let Some((latency_threshold, latency_objective)) = self.latency {
+            objectives.push(Box::new(LatencyObjective {
+                function: self.function,
+                module: self.module,
+                latency_threshold,
+                latency_objective,
+            }));
+        }
+        objectives.into_iter()
+    }
+}
+
+/// Returns the Prometheus recording and alerting rules as a YAML string.
+///
+/// To generate alerts, add the `alerts` parameter to the `autometrics` macro
+/// for at least one function.
+///
+/// Then, call this function to generate the Prometheus rules. You will need
+/// to output the rules to a file and
+/// [load them into Prometheus](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/).
+pub fn generate_alerts() -> String {
+    let groups = METRICS
+        .iter()
+        .flat_map(|alert| {
+            alert.to_objectives().flat_map(|objective| {
+                [
+                    objective.error_ratio_recording_rules(),
+                    objective.meta_recording_rules(),
+                    objective.alert_rules(),
+                ]
+                .into_iter()
+            })
+        })
+        .collect::<Vec<_>>()
+        .join("\n");
+    format!(
+        "---
+# Prometheus recording and alerting rules generated by autometrics-rs
+
+groups:
+{groups}"
+    )
+}
+
+trait Objective {
+    fn slo_type(&self) -> &'static str;
+    fn function(&self) -> &'static str;
+    fn module(&self) -> &'static str;
+    fn success_rate(&self) -> &'static str;
+    fn error_query(&self, window: &str) -> String;
+    fn total_query(&self, window: &str) -> String;
+
+    fn id(&self) -> String {
+        format!("{}-{}", self.module(), self.function())
+    }
+
+    fn filter_labels(&self) -> String {
+        let function = self.function();
+        let module = self.module();
+        let slo_type = self.slo_type();
+        format!("{{function=\"{function}\",module=\"{module}\",objective=\"{slo_type}\"}}")
+    }
+
+    /// When we create a new recording rule, attach these labels to it
+    fn recording_labels(&self) -> String {
```

Comment:
@IvanMerrill @hatchan I'd appreciate it if you guys could look through this file especially and make sure the rules it's generating make sense to you. You can also inspect the output by running the alerts example with `cargo run -p example-alerts generate-alerts > alerting_rules.yml`